# Summary


*   Purpose: The aim of this report is to clean, explore and exam the relationship between different attributes in the student performance data set provide by UCI Machine Learning to predict the impact of family dinamic to student academic performance.

*   Method: The data was provided throught
*   Result
*   Conclusions
*   Recommendation






#Introduction

# Task 1: Data Preparation

This is submission for Assignment 1 for student s3676330

In this notebook, we are going to explore a dataset about student performance from UCI Student Performance dataset, donated by Cortez, Paulo. (2014)

For flexible working enviroment a github repository was set up in order to contain all relevant files as well as the assignment notebook.

The URL to the repo is: https://github.com/Rus268/COSC2999-Assignment1-StudentData.git

In [ ]:
#Task 1: Data Preparation
# "You will start by loading the CSV data from the file (using appropriate pandas functions) and checking whether the loaded data is equivalent to the data in the source CSV file.
# Then, you need to clean the data by using the knowledge we taught in the lectures. You need to deal with all the potential issues/errors in the data appropriately (such as: typos, extra whitespaces, sanity checks for impossible values, and missing values etc). "

# Please structure code as follows:
# always provide one line of comments to explain the purpose of the code, e.g. load the data, checking the equivalent to original data, checking typos (do this for each other types of errors)

## <b> 1.1 Data Import </b>


In [ ]:
# import libraries
import pandas as pd
import numpy as np

In [ ]:
# import data
df = pd.read_csv('https://raw.githubusercontent.com/Rus268/COSC2999-Assignment1-StudentData/main/student_data_csv-Copy1.csv')
# print out data
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  traveltime  studytime  failures  health  absences  G1  \
0    course   mother           2          2         0       3         4   0   
1    course   father           1          2         0       3         2   9   
2     other   mother           1          2         0       3         6  12   
3      home   mother           1          3         0       5         0  14   
4      home   father           1          2         0       5         0  11   
..      ...      ...         ...        ...       ...     ...       ...  ..   
644  course   mother           1          3         1       5         4  10   
645  course   mother           1          2         0       1         4  15   
646  course   mother           2          2         0       5         6  11   
647  course   mother           2          1         0       2         6  10   
648  course   mother           3          1         0       5         4  10   

       G2  G3  
0    11.0  11  
1    11.0  11  
2    13.0  12  
3    14.0  14  
4    13.0  13  
..    ...  ..  
644  11.0  10  
645  15.0  16  
646  12.0   9  
647  10.0  10  
648  11.0  11  

[649 rows x 20 columns]

## <b> 1.2 Basic data description </b> </p>
We can see the data has 20 columns. We have 649 samples in this dataset.

Each sample represent an student with each column represent an attributes of each student. The data was created based on the Student performance dataset donated by Cortez, Paulo (2014). Thus based on the dataset description we can organise all of the attributes as the following table.

|No| Attribute | Expected Value             | Data Category|
|--| ----------| ---------------------------|--------------|
| 1| school    | "GP"(Gabriel Perira) or "MS" (Mousinho da Silveria)|  Binary|         
| 2| sex       | "M"(Male) or "F"(Female)   | Binary       |
| 3| age       | Numeric from 15 to 22      | Interger     |
| 4|address    | "U"(Urban) or "R"(Rural)   | Binary       |
| 5| famsize   | "GT3"(Greater than 3) or LE3(Less or equal to 3)| Binary|
| 6| Pstatus   | "A"(Apart) or "T"(Together)| Binary       |
| 7| Medu      |0-None, 1-primary education(4th Grade), 2-5th to 6th Grade, 3-secondary education or 4-higher education| Interger
| 8| Fedu      |0-None, 1-primary education(4th Grade), 2-5th to 6th Grade, 3-secondary education or 4-higher education| Interger
| 9| Mjob      |"teacher","health" care related, civil "services"(eg Administrive or Police), "at_home" or "other"| Nominal     |
|10| Fjob      |"teacher","health" care related, civil "services"(eg Administrive or Police), "at_home" or "other"| Nominal     |
|11| reason    |close to "home", school "reputation", "course" preference,"other"                           | Nominal     |
|12| guardian  |"mother","father",or "other" | Nominal     |
|13| traveltime| 1-less than 15 min, 2-15 to 30 min, 3- 30 min to 1 hour, 4 - more than 1 hour                             | Ordinal     |
|14| studytime | 1-less than 2 hours, 2-2 to 5 hours, 3-5 to 10 hours, 4-more than 10                                      | Ordinal     |
|15| failures  |


        
</table>




In [5]:
# basic stats of the data
df.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean    17.104777    2.516179    2.305085    1.571649    1.930663    0.223421   
std      7.842476    1.134531    1.098254    0.750430    0.829510    0.593956   
min      0.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    1.000000    1.000000    1.000000    0.000000   
50%     17.000000    2.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max    160.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           health    absences          G2          G3  
count  649.000000  649.000000  648.000000  649.000000  
mean     3.533128    4.118644   11.478395   11.875193  
std      1.445266    8.281493    3.020716    3.263833  
min      1.000000    0.000000    0.000000    0.000000  
25%      2.000000    0.000000   10.000000   10.000000  
50%      4.000000    2.000000   11.000000   12.000000  
75%      5.000000    6.000000   13.000000   14.000000  
max      5.000000  140.000000   19.000000   19.000000

We can see the min age is 0, we expected there might be data error. To further check on what type of error exit, we will attempt to check on the available values in each attributes.

## 1.3 Sanity Check

In [ ]:
# check na
df.isna().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          1
Fjob          1
reason        2
guardian      0
traveltime    0
studytime     0
failures      0
health        0
absences      0
G1            1
G2            1
G3            0
dtype: int64

From NaN check we can see that there are five attributes with empty value:

<ul>
    <li> Mjob (1 Value)
    <li> Fjob (1 Value)
    <li> reason (2 Values)
    <li> G1 (1 Values)
    <li> G2 (1 Values)
</ul>

We will proceed to check the location of the NaN value in order to determine the course of action.

In [ ]:
# Check the location of NaN value in the dataframe.
df[df.isna().any(axis=1)]

school sex  age address famsize Pstatus  Medu  Fedu   Mjob   Fjob reason  \
208     GP   M   16       U      BT       Y     2     3    NaN  other    NaN   
209     GP   M   16       U      GT       Y     3     3  other    NaN    NaN   

    guardian  traveltime  studytime  failures  health  absences   G1   G2  G3  
208       fa           2          1         0       3         0  NaN  3.0   8  
209       mo           3          2         1       3         0    4  NaN   0

We can see that the data are all gather under two observation id 208 and 208. In addition there is also input error that can be observe in famsize, guardian and G2

We will be deleting these two observation as their are only 2 out of 469 observations (

In [ ]:
# Create a Series of unique value for each column
att_value = pd.Series({col:df[col].unique() for col in df.columns})
# Display the Series
att_value

school                                                 [GP, MS]
sex                                                      [F, M]
age           [18, 17, 15, 16, 150, 160, 0, 10, 11, 12, 19, ...
address                                                  [U, R]
famsize                                      [GT3, LE3, BT, GT]
Pstatus                                         [A, T, F, L, Y]
Medu                                            [4, 1, 3, 2, 0]
Fedu                                            [4, 1, 2, 3, 0]
Mjob          [at_home, health, other, services, teacher, se...
Fjob          [teacher, other, services, health, at_home, na...
reason                   [course, other, home, reputation, nan]
guardian      [mother, father, other, futher, mom, momm, fa,...
traveltime                                         [2, 1, 3, 4]
studytime                                          [2, 3, 1, 4]
failures                                           [0, 3, 1, 2]
health                                  

<b>Intitution</b>

Some identifiable mistake include:
<ul>
    <li> Age have value outside of 15 to 22
    <li> famesize have unusual value BT and GT
    <li> Pstatus havve unusual value F, L, Y
    <li> There are NaN value, repeated, miss spelt, and inconsistent data in Fjob and Mjob
    <li> Missing value in reason (nan)
    <li> Input error in guardian
    <li> Absences have value outside of expected range
    <li> 0 is not an expected value for failure.
    <li> G2 value is float instead of Interger
</ul>

We will start by further examinating the column with clear error and devices a potential solution to fix it.

## 1.4 Cleaning Data Error

In [ ]:
# Check value of age:
df["age"].value_counts()

age
17     178
16     175
18     140
15     106
19      32
20       7
0        2
10       2
21       2
150      1
160      1
11       1
12       1
22       1
Name: count, dtype: int64

From value check for age we can see that there are impossible value such as 0, 150 and 160. The expected range of the data is 15 to 22.

Propose solution:
- Modify the 150 to 15
- Modify 160 to 16
- Replace outside range value 0, 10, 11, 12 with 15

In [ ]:
# Replace 150 to 15
df["age"].replace(150,15, inplace = True)
# Replace 160 to 16
df["age"].replace(160,16, inplace = True)
# Replace [0, 10, 11, 12] to 15
df["age"].replace([0,10,11,12],15, inplace = True)
# Check the value of age again
df["age"].value_counts()

17    178
16    176
18    140
15    113
19     32
20      7
21      2
22      1
Name: age, dtype: int64

In [ ]:
# Check famsize value
df["famsize"].value_counts()

famsize
GT3    456
LE3    191
BT       1
GT       1
Name: count, dtype: int64

Expect only Binary response either GT3 or LE3. Thus BT and GT are outlier.

As G and B are close to each other on the keyboard and BT and GT are both close to GT3 thus it is suspected that these are miss input for GT3

Propose solution is changing GT and BT to GT3

In [ ]:
# Change "BT","GT" to "GT3" in the data frame
df["famsize"].replace(["BT","GT"],"GT3", inplace = True)
# Display value of famesize
df["famsize"].value_counts()

GT3    458
LE3    191
Name: famsize, dtype: int64

In [ ]:
# Check Pstatus
df["Pstatus"].value_counts()

Pstatus
T    523
A     79
F     42
Y      3
L      2
Name: count, dtype: int64

Expected value only T for "Together" or A for "Apart". Thus unexpected value are F (42), Y (3), L (2). Since "Y","F" are close to "T" it is highly likely that these are miss input in the attempt of typing in "T". As for "L"

In [ ]:
# Check Mjob
df["Mjob"].value_counts(dropna = False)

Mjob
other       258
services    132
at_home     128
teacher      72
health       48
at home       4
service       3
home          3
NaN           1
Name: count, dtype: int64

As the NaN check show that Mjob have one value of NaN, we used the command value_counts(dropna = False) to include NaN value for checking purpose.

We can see that there is:
<ul style="list-style-type:circle">
    <li>blank space error (<b>at home</b> instead of <b>at_home</b>)
    <li>wrong value (<b>home</b> instead of <b>at_home</b>)
    <li>grammar error (<b>service</b> instead of <b>services</b>
    <li>Missing value (NaN)
</ul>

Propose fix:
- replace "at home", "home" to "at_home"
- replace "service" to "services"
- replace NaN to "other" the most common value which would cover other situation.

In [ ]:
# Replace "at home","home" to "at_home"

# Replace "service" to "services"


In [ ]:
# Check reason:
df["reason"].value_counts(dropna=False)

reason
course        284
home          148
reputation    143
other          72
NaN             2
Name: count, dtype: int64

In [ ]:
# Check guardian:
df["guardian"].value_counts(dropna=False)

guardian
mother    443
father    146
other      41
mom        11
futher      5
momm        1
fa          1
mo          1
Name: count, dtype: int64

Propose fixed:
- Standardize "mom", "momm", "mo" to "mother"
- Standardize "futher", "fa" to "father"

In [ ]:
# Replace ["mom", "momm", "mo"] to "mother"

# Replace ["futher","fa"] to "father"


In [ ]:
# Check absences
df["absences"].value_counts()

absences
0      245
2      108
4       93
6       49
8       39
10      21
12      12
5       12
1       12
16      10
9        7
3        7
14       7
11       5
18       3
7        3
20       2
22       2
21       2
15       2
80       2
32       1
30       1
140      1
26       1
13       1
24       1
Name: count, dtype: int64

There is some outlier data in this case is 140. This data is outside of the expected scope thus it very likely it is an error.

In [ ]:
# Check G1
df["G1"].value_counts()

G1
10     93
11     90
12     81
14     70
13     70
9      65
8      42
15     35
7      33
16     22
17     16
6       9
18      7
5       5
X       3
4       3
0       1
132     1
123     1
19      1
Name: count, dtype: int64

Value that need to be fix:
<ul>
    <li> X value likely to be wrong. Most likely it is 10
    <li> Outside of expected range 132, 123. Bring it back to the closes value 12,13
</ul>

In [ ]:
# Check G2
df["G2"].value_counts(dropna=False)

11.0    103
12.0     85
10.0     83
13.0     80
9.0      72
14.0     53
8.0      40
15.0     37
16.0     24
17.0     19
7.0      16
18.0     14
0.0      10
6.0       7
5.0       3
3.0       1
NaN       1
19.0      1
Name: G2, dtype: int64

As this is inconsistent with G1 and G3 we will be removing the zero after the comma by rounding it up to 0 decima

In [ ]:
# Rounding the G2 value to 0 decima


TypeError: ignored

In [ ]:
# Check G3
df["G3"].value_counts(dropna=False)

11    104
10     97
13     82
12     71
14     63
15     49
8      36
9      35
16     35
17     29
0      16
18     15
7      10
6       3
19      2
1       1
5       1
Name: G3, dtype: int64

Finally we export the dataframe which contain the cleaned data into a new CSV files for the next exploration steps.

In [ ]:
# Export the newly clean data into a new CVS file
# df.to_csv("student_data_cvs_clean.cvs")

# Task 2: Data Exploration

In [ ]:
# Task 2: Data Exploration
# 1. Choose 1 column with nominal values, 1 column with ordinal Values, and 1 column with numerical values
df

#Code goes after this line



school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  traveltime  studytime  failures  health  absences  G1  \
0    course   mother           2          2         0       3         4   0   
1    course   father           1          2         0       3         2   9   
2     other   mother           1          2         0       3         6  12   
3      home   mother           1          3         0       5         0  14   
4      home   father           1          2         0       5         0  11   
..      ...      ...         ...        ...       ...     ...       ...  ..   
644  course   mother           1          3         1       5         4  10   
645  course   mother           1          2         0       1         4  15   
646  course   mother           2          2         0       5         6  11   
647  course   mother           2          1         0       2         6  10   
648  course   mother           3          1         0       5         4  10   

       G2  G3  
0    11.0  11  
1    11.0  11  
2    13.0  12  
3    14.0  14  
4    13.0  13  
..    ...  ..  
644  11.0  10  
645  15.0  16  
646  12.0   9  
647  10.0  10  
648  11.0  11  

[649 rows x 20 columns]

In [ ]:
# Task 2: Data Exploration
# 2. Explore the relationships between columns; at least 3 visualisations with plausible hypothesis

#Code goes after this line




In this section we will be exploring the clean dataset.

Selected column are:
- G3 (Final grade) - numerial
- guardian - nominal value
- Fedu - ordinal values

Relationship hypothesis:
- There is a correlation between guardian education level and student final grade.


In [ ]:
# Task 2: Data Exploration
# 3. Scatter matrix for all numerical columns

#Code goes after this line

# Task 3: Dealing with Missing Value and Outlier

# Task 4: External Data